test/method.jl

```diff
@@ -0,0 +1,182 @@
+macro test_signature(function_def_expr, method=nothing)
+    _target = splitdef(function_def_expr)
+    return quote
+        fun = $(esc(function_def_expr))
+        meth = if ($method === nothing)
+            only_method(fun)
+        else
+            $(esc(method))
+        end
+        sig = signature(meth)
+        test_matches(sig, $(_target))
+    end
+end
+
+function test_matches(candidate::AbstractDict, target::Dict)
+    # we want to use literals in the tests so that @test gives useful output on failure
+    haskey(target, :name) && @test target[:name] == get(candidate, :name, nothing)
+    haskey(target, :params) && @test target[:params] == get(candidate, :params, nothing)
+    haskey(target, :args) && @test target[:args] == get(candidate, :args, nothing)
+    haskey(target, :whereparams) &&
+        @test target[:whereparams] == get(candidate, :whereparams, nothing)
+    return nothing
+end
+
+"""
+    only_method(f, [typ])
+
+Return the only method of `f`,
+filtering by those accepts the types indicated by `typ` if it is provided.
```

Comment:
Could this be made clearer? Not sure what it is saying

Reply:
Can we just get rid of this line?
The next line should cover it?

Reply:
yeah, i think this line can go
```suggestion
```

Reply:
it might be worth documenting it'll throw if `f` has multiple or zero methods?

also fwiw `only(methods(f))` would throw a `ArgumentError` rather than an `ErrorException`

Reply:
also if this is just `only(methods(f))` maybe just depend on Compat?